### DISCUSSION TITLE
Unable to install a library (pact-python) which depends on fastapi because of starlette version conflict

### DISCUSSION THREAD

**Discussion by meadsteve** [fastapi/fastapi#8783] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# requirements.txt

starlette==0.16.0

# This next one depends on fastapi
pact-python==1.4.0
```

### Description

pip install -r requirements.txt

### Operating System

macOS

### FastAPI Version

0.67.0 

### Python Version

3.9

### Additional Context

One fix I can think of is to make fastapi depend not on a fixed version of starlette `starlette ==0.14.2"` but on a range maybe`starlette ~=0.14.2",`. I wanted to open an issue first to discuss this as it has testing/release implications but I'm happy to investigate a PR if you think this is a good idea.

Was previously an issue with https://github.com/tiangolo/fastapi/issues/1407

**Comment by Kludex** [comment]
Already being implemented on https://github.com/tiangolo/fastapi/pull/3594

A version range is not possible.

**Comment by meadsteve** [comment]
@Kludex What are the problems with the range? Maybe that's something we could look into. Obviously the fix in #3594 is great but I wonder if this will just keep happening

**Comment by Kludex** [comment]
The problem is the breaking changes introduced by the new version.

It will keep happening as Starlette is not stable.

**Comment by adriangb** [comment]
It might be beneficial for FastAPI to mark itself as "unstable" then and just accept (and integrate) any breaking changes coming from its dependencies. IMO everyone should have pinned dependencies in their deployables, so it shouldn't be a problem for services/apps using FastAPI. It does generate more work for downstream _libraries_ though.

A strong argument for this is: what happens if a critical vulnerability is found in Starlette right now? FastAPI is pinned to `0.14`, but Starlette might release the fix in `0.16.2`. Either FastAPI has to rush to bump the Starlette version, do a bunch of internal juggling to manage the breaking changes and release its own breaking change, or Starlette has to backport. This creates a situation where users have to introduce breaking changes just to get a security fix.

**Comment by graingert** [comment]
https://github.com/getslash/pact/blob/develop/setup.cfg pact doesn't seem to depend on FastAPI at all

**Comment by graingert** [comment]
```bash
$ cat requirements.in 
pact>=1.4.0
starlette>=0.16.0
$ cat requirements.txt 
#
# This file is autogenerated by pip-compile with python 3.9
# To update, run:
#
#    pip-compile --allow-unsafe --generate-hashes
#
anyio==3.3.0 \
    --hash=sha256:929a6852074397afe1d989002aa96d457e3e1e5441357c60d03e7eea0e65e1b0 \
    --hash=sha256:ae57a67583e5ff8b4af47666ff5651c3732d45fd26c929253748e796af860374
    # via starlette
flux==1.3.5 \
    --hash=sha256:6e2bf372c69c1c2f77b043d89a72cf7b6a53fbd91bdda9210fc5da085e3e61ab
    # via pact
idna==3.2 \
    --hash=sha256:14475042e284991034cb48e06f6851428fb14c4dc953acd9be9a5e95c7b6dd7a \
    --hash=sha256:467fbad99067910785144ce333826c71fb0e63a425657295239737f7ecd125f3
    # via anyio
logbook==1.5.3 \
    --hash=sha256:0cf2cdbfb65a03b5987d19109dacad13417809dcf697f66e1a7084fb21744ea9 \
    --hash=sha256:2dc85f1510533fddb481e97677bb7bca913560862734c0b3b289bfed04f78c92 \
    --hash=sha256:56ee54c11df3377314cedcd6507638f015b4b88c0238c2e01b5eb44fd3a6ad1b \
    --hash=sha256:66f454ada0f56eae43066f604a222b09893f98c1adc18df169710761b8f32fe8 \
    --hash=sha256:7c533eb728b3d220b1b5414ba4635292d149d79f74f6973b4aa744c850ca944a \
    --hash=sha256:8f76a2e7b1f72595f753228732f81ce342caf03babc3fed6bbdcf366f2f20f18 \
    --hash=sha256:94e2e11ff3c2304b0d09a36c6208e5ae756eb948b210e5cbd63cd8d27f911542 \
    --hash=sha256:97fee1bd9605f76335b169430ed65e15e457a844b2121bd1d90a08cf7e30aba0 \
    --hash=sha256:e18f7422214b1cf0240c56f884fd9c9b4ff9d0da2eabca9abccba56df7222f66
    # via pact
pact==1.12.0 \
    --hash=sha256:412d09c955ec385e49c1ad73204a5afe92a720cc9134b51c22c20ef17aa14b80
    # via -r requirements.in
sniffio==1.2.0 \
    --hash=sha256:471b71698eac1c2112a40ce2752bb2f4a4814c22a54a3eed3676bc0f5ca9f663 \
    --hash=sha256:c4666eecec1d3f50960c6bdf61ab7bc350648da6c126e3cf6898d8cd4ddcd3de
    # via anyio
starlette==0.16.0 \
    --hash=sha256:38eb24bf705a2c317e15868e384c1b8a12ca396e5a3c3a003db7e667c43f939f \
    --hash=sha256:e1904b5d0007aee24bdd3c43994be9b3b729f4f58e740200de1d623f8c3a8870
    # via -r requirements.in
waiting==1.4.1 \
    --hash=sha256:b641ef3a2382e101d3c4e66494da6fc52b82c325dd04ae341e051b0eec6f30cd
    # via pact
```

**Comment by meadsteve** [comment]
@graingert sorry I should have said pact-python. I forget there's two libraries called pact 🤦 https://pypi.org/project/pact-python/.

(I've updated this issue - though it's not really specific to pact-python that's just how I found it)

**Comment by tiangolo** [comment]
About the original question @meadsteve, you probably shouldn't pin Starlette as part of your dependencies if you are using FastAPI, you should pin FastAPI instead. You can read more about it here: https://fastapi.tiangolo.com/deployment/versions/

You could also use a lock file, e.g. with `pip-tools` or Poetry, but that would be generated automatically and depend on the original requirements of your app, e.g. FastAPI.

**Comment by meadsteve** [comment]
@tiangolo my situation was a little complicated. In this case the project I was working on used Starlette itself. But it also used pact which depended on fastapi with a conflicting Starlette. So we *had* to pin Starlette. 

So I'm not really sure what the best thing to do here is. It's mostly just python packaging can be complicated 😆
